### PR TITLE
New authentication for API

### PR DESCRIPTION
**Pull request by mishuagopian**
## Summary

* Changed to the authentication system that is specified in WBooks a apiary.

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -36,7 +36,40 @@ exports.login = (req, res, next) => {
...
+  const user = req.user;
+  const SALT_ROUNDS = 10;
+
+  bcrypt.genSalt(SALT_ROUNDS).then((verificationCode) => {
+    user.verificationCode = verificationCode;
+    user.save();
+    res.status(200);
+    res.end();
+  }).catch(() => {
+    res.status(500);
```
Comment (sbalay) [inline app/controllers/users.js]: shouldn't you just throw an exception here?
That way the response will still be 500, and then you can read the logs. This way no logs are generated

File: app/controllers/users.js [hunk 2]
```diff
@@ -59,42 +92,49 @@ exports.update = (req, res, next) => {
...
+    }).catch((err) => {
+      res.status(400);
+      res.send({ error: 'Invalid password' });
     });
-  }).catch((err) => {
-    res.status(400);
-    res.send({ error: 'Invalid password' });
+
+  }).catch(() => {
+    res.status(500);
```
Comment (sbalay) [inline app/controllers/users.js]: same here

File: app/routes.js [hunk 3]
```diff
@@ -7,6 +7,10 @@ exports.init = (app) => {
 
   // Users
   app.post('/users/sessions', [], users.login);
+
+  app.post('/users/sessions/renew', [m.secure], users.renew);
+  app.post('/users/sessions/invalidate', [m.secure], users.invalidateAll);
```
Comment (epintos) [inline app/routes.js]: invalidate_all ?

File: test/users.js [hunk 4]
```diff
@@ -44,17 +44,76 @@ describe('users', () => {
...
+        .end((err, res) => {
+          res.should.have.status(401);
+          done();
+        });
+    });
+
+    it('should be successful', (done) => {
+      successfulLogin((loginErr, loginRes) => {
+        chai.request(server)
+          .post('/users/sessions/invalidate')
```
Comment (epintos) [inline test/users.js]: invalidate_all ?

File: app/controllers/users.js [hunk 5]
```diff
@@ -36,7 +36,40 @@ exports.login = (req, res, next) => {
...
+  const accessToken = req.accessToken;
+  const user = req.user;
+
+  if (req.body.renew_id === accessToken.renewId) {
+    if (moment().isBefore(accessToken.expirationDateWarning)) {
+      res.status(403);
+      res.send({ error: 'Warning expiration date has not been reached' });
+    } else {
+      res.status(200);
+      res.send({ access_token: sessionManager.generateAccessTokenWithRenewId(user, accessToken.renewId) });
```
Comment (mpizzagalli) [inline app/controllers/users.js]: the `generateAccesTokenWithRenewId` calculates each time the `maximumUsefulDate`, but here we shouldn't be updating that value, or am I missing something?

Comment (mishuagopian) [inline app/controllers/users.js]: Yes, you are right. That's a bug I had already found but as we are not merging this PR I never fixed it. Ty for the review!

**Pull request closed without merging**